lib/galaxy/tools/evaluation.py

```diff
@@ -634,11 +634,28 @@ def _build_environment_variables(self):
             if inject == "api_key":
                 if self._user and isinstance(self.app, BasicSharedApp):
                     from galaxy.managers import api_keys
-
                     environment_variable_template = api_keys.ApiKeyManager(self.app).get_or_create_api_key(self._user)
                 else:
                     environment_variable_template = ""
                 is_template = False
+            elif inject == "oidc_refresh_token":
```

Comment:
There could be an arbitrarily long wait time until the job starts and the values are actually used, is that going to be a problem ?

Reply:
I think users can be logged in  / associated with multiple providers, do we need to anticipate the possibility that tools require "provider-specific" (google, azure, some other more specific service) tokens ?

Reply:
> There could be an arbitrarily long wait time until the job starts and the values are actually used, is that going to be a problem ?

It depends, this is actually an issue for `access_token` and `id_token`, this is why we also expose `refresh_token`, which usually has a much longer expiration time (24+ hours), and it depends on the identity provider. If a job starts after that, there is hardly much we can do - single sign-on would not work. I was thinking about adding a Galaxy API to get a fresh token, but we have to see if there are security considerations and really see the use case where a job start takes that long.

Reply:
> I think users can be logged in / associated with multiple providers, do we need to anticipate the possibility that tools require "provider-specific" (google, azure, some other more specific service) tokens ?

This is a good catch, thank you! I'm actually always using the first token,
https://github.com/galaxyproject/galaxy/pull/15300/files#diff-59a534d07c0974fda8548d41482653dd5512d69681950449207ec5c0f1ccc3f3R757
but if there are more than one, that might not work. We'd need to get the one from the provider that was used to login.

Reply:
@mvdbeek - I do not see a nice way to get information about the provider used for login. I think it is not stored in Galaxy, but please correct me if I'm wrong. If I'm not - one could change a database scheme and add a field to a user model, say `last_used_idp` and set it during login. I'm a bit reluctant to change the database scheme until someone confirms that there is no other way. And that could be a new PR, to keep them manageable.

Reply:
xref https://github.com/galaxyproject/galaxy/issues/14578, yes, we absolutely need to specify the provider. There I proposed a tool syntax like:

```
    <environment_variables>
        <environment_variable name="LS_TOKEN" provider="elixir" inject="access_token" />
    </environment_variables>
```

since in many cases the tool is specific to one endpoint and will only accept one flavour of token.

Reply:
Hm, we've been thinking about the situation where Galaxy is configured to work with multiple providers. For us, it is kind of an edge case when the same user logs in using multiple providers. Why would that happen? And what is the expected behavior? Should we automatically update all tokens? What should we do if one of the updates fails, but another is ok - logout user or not?
So, what I currently implemented - if a user logs in using one provider, tokens from all other providers are invalidated. So there can only be one set of active tokens from a specific provider. Example - a user logs in using Elixir, then the same user logs in using Google - Elixir tokens will be deleted, only Google ones will be refreshed.

As for your @hexylena idea to specify the provider in a tool - this is a good one! But I'd suggest making it optional, because some tools might just need a token (to pass it further to some service) and mustn't know what kind of token it is. So, if provider is specified - we check if such token is active and return it, if not - we just return an active token.


Does this all make sense?

Reply:
> For us, it is kind of an edge case when the same user logs in using multiple providers.

I somewhat agree here, it's probably relatively rare.

> Why would that happen? And what is the expected behavior? Should we automatically update all tokens? What should we do if one of the updates fails, but another is ok - logout user or not?

I guess the question is: how common is the case where a tool needs a specific token? (I guess this will support two cases: tools and built-in functionality, but focusing on tools for a moment.)

For the tools I have in mind they all need a specific provider, and if another provider token update fails that's absolutely fine. I don't think token refreshing should log out the user immediately, I don't know if that's a valid thing to do (does a failed refresh strictly imply loss of access to the account? e.g. in a temporary backend failure it wouldn't make sense to kick the user.)

For the built-in functionality case like accessing remote data / storing datasets to a user provided backend, then, yeah, maybe makes a lot more sense to kick them immediately since everything will break.

> So, what I currently implemented - if a user logs in using one provider, tokens from all other providers are invalidated. So there can only be one set of active tokens from a specific provider. Example - a user logs in using Elixir, then the same user logs in using Google - Elixir tokens will be deleted, only Google ones will be refreshed.

Interesting! I can see why, but, yeah for the specific elixir case, there's a tool the user wants to execute which requires elixir authentication. It would be absolutely fine then to e.g. store that data on google or mix that data with data from google, and having them go through a logout/login flow just to execute a tool seems rather unpleasant for the user?

> But I'd suggest making it optional, because some tools might just need a token (to pass it further to some service) and mustn't know what kind of token it is.

yes, making it optional is fine for me.

(but personally I'm curious what kind of tools can use a token completely interchangeably, wouldn't this also require selecting different API endpoints or so?)

Reply:
> , I don't know if that's a valid thing to do (does a failed refresh strictly imply loss of access to the account? e.g. in a temporary backend failure it wouldn't make sense to kick the user.)

I think we should kick the user. If one logged in through some provider and we cannot refresh (read re-verify) the user, we cannot trust this user anymore. So, I think we should log them out. One thing we can do is to add a grace period -  since we refresh tokens before the access token expires anyway, we can return ok in case of failure for some time and only return error after the original access_token expires.


> It would be absolutely fine then to e.g. store that data on google or mix that data with data from google, and having them go through a logout/login flow just to execute a tool seems rather unpleasant for the user?

I'm not sure I quite understand that. So a user logs in via Google, does something with data, then logs out, and logs in with Elixir. But, in spite of logout, do we still want to keep Google tokens and refresh them? And if that Google account does not exist, are we still ok?

I can try to implement that (as I said, for us that should never happen, so we should be good) if everybody agrees. In this case, we'd be refreshing all tokens, never logout a user if refresh fails, and asking for a provider in a tool. 


> (but personally I'm curious what kind of tools can use a token completely interchangeably, wouldn't this also require selecting different API endpoints or so?)

we have such a tool - it grabs some data via scp, using OIDC token to login - so no need to know what this token is, we just pass it to sshd service on the other side.


Btw, custos auth does not allow multiple providers (so you cannot have Keycloak and Google together for the same user).

Reply:
> I think we should kick the user. If one logged in through some provider and we cannot refresh (read re-verify) the user, we cannot trust this user anymore. So, I think we should log them out. One thing we can do is to add a grace period - since we refresh tokens before the access token expires anyway, we can return ok in case of failure for some time and only return error after the original access_token expires.

Yeah ok. I can agree to it. I'm worried about edge cases, but, maybe they're just that: edge cases.

> I'm not sure I quite understand that. So a user logs in via Google, does something with data, then logs out, and logs in with Elixir. But, in spite of logout, do we still want to keep Google tokens and refresh them? And if that Google account does not exist, are we still ok?

Yeah ok, maybe again I'm too worried about multi-auth'd accounts. I know we'll have users that login via elixir and never use their token, and other users that login explicitly to use that token in the tool which consumes it.

I guess I'm working under the impression that by also having an e.g. Google account linked, in addition to their elixir account, that they can do different things with that. That they'll never explicitly click "logout", but instead:

- login with elixir
- attach google as well via user preferences (rather than logging out and logging in with google which would generate a second galaxy account because they don't resolve the same way)
- do neat things that require tokens from both

> In this case, we'd be refreshing all tokens, never logout a user if refresh fails, and asking for a provider in a tool.

yeah, in **this** case _not_ kicking the user makes more sense.
But if they only logged in via google, and the token fails to validate, then I agree we should kick them.

For the _vast majority_ of our users, they had an existing account, and they attached a social login to that to get some benefit, rather than primarily logging in with social auth. 

But for your case where they've only ever logged in via social, hard agree, kick if validation fails. But those users won't have proper passwords right? So many the logic comes down to, in the event of validation failure

- if only one social
  - and no password ⇒ kick
  - and a password ⇒ they can still login, don't kick
- if multiple socials
  - and they all fail to validate ⇒ kick
  - and they have a password ⇒ don't kick.

> we have such a tool - it grabs some data via scp, using OIDC token to login - so no need to know what this token is, we just pass it to sshd service on the other side.

huh, neat!

> Btw, custos auth does not allow multiple providers (so you cannot have Keycloak and Google together for the same user).

good reason for me to not switch to custos :)

Reply:
ok, I think I start to understand :). Thank you for the clarifications!

So, we basically have two scenarios here. The first one is when a user logs in using social auth. In this case one would not normally expect tokens from multiple providers and would expect token verification at each request and logout if something is wrong. This is the scenario I had in mind.

A second scenario is when an existing user connects to a third-party IDP using user preferences. I was not even aware of that option before :(, but I guess this is more your use case. In this case you indeed do not want to logout user and you do want to select which token to expose in tools.

Ideally, we would distinguish these two scenarios, keeping track of which provider a user used to log in and which providers they connected extra. Then we could return the "provider-last-used-to-login" token as default, kick off a user only if the "provider-last-used-to-login" token cannot be refreshed (if there is such a token), and refresh all other tokens.
Should I go forward and implement that? It would require adding a field to the user table in the database, which I tried to avoid before.

Alternatively, as I said before, we can just refresh everything and never kick users. Worst case - we allow a user to be logged in with an expired token. It would work for us since they would not be able to access data/resources.

Reply:
> A second scenario is when an existing user connects to a third-party IDP using user preferences. I was not even aware of that option before :(, but I guess this is more your use case. In this case you indeed do not want to logout user and you do want to select which token to expose in tools.

> Ideally, we would distinguish these two scenarios, keeping track of which provider a user used to log in and which providers they connected extra. Then we could return the "provider-last-used-to-login" token as default, kick off a user only if the "provider-last-used-to-login" token cannot be refreshed (if there is such a token), and refresh all other tokens. Should I go forward and implement that? It would require adding a field to the user table in the database, which I tried to avoid before.

Would an explicit credential selector input that gets rendered to the tool form resolve this?  In the first situation, with only one credential, it wouldn't allow you to change anything but it would make it clear to the user that a tool would be using their creds.  In the second situation, you could provide the specific credential set to use, toggling between google and azure, or whatever you may have.

Reply:
> Would an explicit credential selector input that gets rendered to the tool form resolve this?

Can we dynamically build an input list for this selector? The values should depend on the currently attached identities for a specific user. I'm not sure how to implement this. Could you please give me a hint?